openstack/blockstorage/extensions/volumeactions/requests.go

```diff
@@ -267,3 +267,34 @@ func ForceDelete(client *gophercloud.ServiceClient, id string) (r ForceDeleteRes
 	_, r.Err = client.Post(actionURL(client, id), map[string]interface{}{"os-force_delete": ""}, nil, nil)
 	return
 }
+
+// ImageMetadataOptsBuilder allows extensions to add additional parameters to the
+// ImageMetadataRequest request.
+type ImageMetadataOptsBuilder interface {
+	ToImageMetadataMap() (map[string]interface{}, error)
+}
+
+// ImageMetadataOpts contains options for setting image metadata to a volume.
+type ImageMetadataOpts struct {
+	// The image metadata to add to the volume as a set of metadata key and value pairs.
+	Metadata map[string]string `json:"metadata"`
+}
+
+// ToImageMetadataMap assembles a request body based on the contents of a
+// ImageMetadataOpts.
+func (opts ImageMetadataOpts) ToImageMetadataMap() (map[string]interface{}, error) {
+	return gophercloud.BuildRequestBody(opts, "os-set_image_metadata")
+}
+
+// ImageMetadata will set image metadata on a volume based on the values in ImageMetadataOptsBuilder.
+func ImageMetadata(client *gophercloud.ServiceClient, id string, opts ImageMetadataOptsBuilder) (r ImageMetadataResult) {
```

Comment:
Sorry, I had meant for this to stay `SetImageMetadata` (https://github.com/gophercloud/gophercloud/pull/1621#discussion_r293184159)

The idea is that the opts structure is a "thing" and the function is an "action". We're setting image metadata with the options defined in the image metadata struct, if that makes sense.

Same with the result - that can stay as `SetImageMetadataResult` because it's the result of the action.